### ISSUE TITLE
Enable movement to any valid square within player range

### ISSUE THREAD

**Issue by p-d-moore** [njustesen/ffai#75]
Currently in the client, each move must be manually broken into a series of individual 1 square steps.  I propose enabling the client to graphically show all available squares a player can move to, and allow any of those squares to be clicked on (or selected by a bot).  This will dramatically improve the use of the interface and additionally simplify the implementation of AI bots.

In alternative clients such as FUMBBL and Cyanide, when a human player selects a move (or blitz) action for a unit, all available squares to move to are highlighted and colour coded for the probability of reaching that square succesfully.  Clicking on any square causes the unit to traverse a minimum probability cost sequence of steps to that square, generally with the lowest used movement for that probability cost.

The implementation of this in the current client would involve the application of the pathfinding.py module, and some graphical and code changes to queue up a sequence of moves.  The queuing of sequence of moves is something the GrodBot already does, as a means of inspiration.

**Comment by njustesen** [comment]
This would be very nice. 

I suppose we need function that returns all squares a player can reach, and for each square, the path to it with the dice rolls to make for each step. Would it be fast enough?

I can easily do the user interface when the pathfinding part is working.

**Comment by p-d-moore** [comment]
Yes it would be fast enough, as it only need to calculate for the player / action selected (as opposedto every player and every action as the Grodbot currently does).  I can write the function easily enough, it's just cobbling together code I already wrote.

**Comment by njustesen** [comment]
Any progress on this @p-d-moore?

**Comment by mrbermell** [comment]
I've added this feature. Runnable code in my personal branch [https://github.com/mrbermell/ffai/tree/all-paths-in-available-actions](https://github.com/mrbermell/ffai/tree/all-paths-in-available-actions)

In summary I added three options which can be set by the game config. 
**No roll paths** - Only show squares that a player can reach without rolling any dice (except for the adjacent squares).  
**Single rolls paths** - As above but add one more square i.e. the dodge and/or the GFI. 
**Full paths** - all GFIs and dodges. All the rolls for one particular path are shown in the destination square (bloaty!). 
![image](https://user-images.githubusercontent.com/31434035/101224506-15dd9b00-368f-11eb-93ba-e13d4da513d3.png)
 
_No roll_ and _single roll paths_ are calculated noticeably faster than the full paths, especially near multiple opponents. My parts of the code can probably be more efficient... 

I'm not sure if I implemented the configuration a neat way. I ponder whether an Agent should be able to say what kind of pathfinding it prefers. 

Finally, I welcome all feedback and ideas!

**Comment by njustesen** [comment]
Very nice! I'll check it out. 

Looks like the UI needs to be fine-tuned a bit :)

**Comment by mrbermell** [comment]
Yeah, showing one path and its corresponding rolls when hovering a possible square with the pointer comes to mind. Though I imagine that's quite an effort. 

Regarding the options for agents. I believe it's necessary to add a flag in the agent for pathfinding. And only calculate the paths if Game.actor is human or has the flag set. The scripted bots and random bots don't need it, only the ML bots do. I think adding a function in Game called next_actor_wants_pathfinding() is reasonable. 

I don't like the procedures accessing the agent directly because of old OOP principles, hence the awkwardly named function.

**Comment by mrbermell** [comment]
I've modified the Single Roll Path option to be all paths that only contain a single roll but the square with the roll doesn't need to be the last. I player a game with this setting and though you feel really lazy it's feels really good. 

I also added some logic for ending a player's turn after movement under some conditions to speed up ML training. 

Below picture shows an example, the active player can dodge in to its opponent's half and the 3+dice rolls that are shown are from that initial dodge. Once you get used to player with it, it's nice. 
![image](https://user-images.githubusercontent.com/31434035/102933001-aa266b00-44a1-11eb-980b-9f4dfdd090e2.png)
